designs/event-streams.md

```diff
@@ -0,0 +1,275 @@
+        if self.output_stream is None:
+            _, self.output_stream = await self.await_output()
+
+        await self.input_stream.close()
+        await self.output_stream.close()
+
+    async def __aenter__(self) -> Self:
+        return self
+
+    async def __aexit__(self, exc_type: Any, exc_value: Any, traceback: Any):
+        await self.close()
+
+
+class InputEventStream[I: SerializableShape, R](Protocol):
+
+    input_stream: AsyncEventPublisher[I]
+
+    _response: R | None = None
+
+    @property
+    def response(self) -> R | None:
+        return self._response
+
+    @response.setter
+    def response(self, value: R) -> None:
+        self._response = value
+
+    async def await_output(self) -> R:
+        ...
+
+    async def close(self) -> None:
+        await self.input_stream.close()
+
+    async def __aenter__(self) -> Self:
+        return self
+
+    async def __aexit__(self, exc_type: Any, exc_value: Any, traceback: Any):
+        await self.close()
+
+
+class OutputEventStream[O: DeserializableShape, R](Protocol):
+
+    output_stream: AsyncEventReceiver[O]
+
+    response: R
+
+    async def close(self) -> None:
+        await self.output_stream.close()
+
+    async def __aenter__(self) -> Self:
+        return self
+
+    async def __aexit__(self, exc_type: Any, exc_value: Any, traceback: Any):
+        await self.close()
+```
+
+All three classes share certain functionality. They all implement `close` and
+the async context manager magic methods. By default these just call close on
+the underlying publisher and/or receiver.
+
+Both `InputEventStream` and `DuplexEventStream` have an `await_output` method
+that waits for the initial request to be received, returning that and the output
+stream. Their `response` and `output_stream` properties will not be set until
+then. This is important because clients MUST be able to start sending events to
+the service immediately, without waiting for the initial response. This is
+critical because there are existing services that require one or more events to
+be sent before they start sending responses.
+
+```python
+with await client.duplex_operation(DuplexInput(spam="eggs")) as stream:
+    stream.input_stream.send(FooEvent(foo="bar"))
+
+    initial, output_stream = await stream.await_output()
+
+    for event in output_stream:
+        handle_event(event)
+
+
+with await client.input_operation() as stream:
+    stream.input_stream.send(FooEvent(foo="bar"))
+```
+
+The `OutputEventStream`'s initial `response` and `output_stream` will never be
+`None`, however. Instead, the `ClientProtocol` MUST set values for these when
+constructing the object. This differs from the other stream types because the
+lack of an input stream means that the service has nothing to wait on from the
+client before sending responses.
+
+```python
+with await client.output_operation() as stream:
+    for event in output_stream:
+        handle_event(event)
+```
+
+## FAQ
```

Comment:
A couple of general questions (not sure where they belong)

How are error and unknown events handled?  And how are errors in the initial response (esp for Input/Duplex stream) handled?

How/where is event signing done - is this part of event publishers?